pkg/backup/rbac.go

```diff
@@ -0,0 +1,143 @@
+/*
+Copyright 2018 the Heptio Ark contributors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package backup
+
+import (
+	arkdiscovery "github.com/heptio/ark/pkg/discovery"
+	"github.com/pkg/errors"
+	"github.com/sirupsen/logrus"
+	rbac "k8s.io/api/rbac/v1"
+	rbacbeta "k8s.io/api/rbac/v1beta1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/client-go/kubernetes"
+	rbacclient "k8s.io/client-go/kubernetes/typed/rbac/v1"
+	rbacclientbeta "k8s.io/client-go/kubernetes/typed/rbac/v1beta1"
+)
+
+type ClusterRoleBindingLister interface {
+	// List returns a slice of ClusterRoleBindings which can represent either v1 or v1beta1 ClusterRoleBindings.
+	List() ([]ClusterRoleBinding, error)
+}
+
+type v1ClusterRoleBindingLister struct {
+	client rbacclient.ClusterRoleBindingInterface
+}
+
+func (v1 v1ClusterRoleBindingLister) List() ([]ClusterRoleBinding, error) {
+	crbList, err := v1.client.List(metav1.ListOptions{})
+	if err != nil {
+		return nil, errors.WithStack(err)
+	}
+	var crbs []ClusterRoleBinding
+	for _, crb := range crbList.Items {
+		crbs = append(crbs, v1clusterrolebinding{crb: crb})
+	}
+
+	return crbs, nil
+}
+
+type v1beta1ClusterRoleBindingLister struct {
+	client rbacclientbeta.ClusterRoleBindingInterface
+}
+
+func (v1beta1 v1beta1ClusterRoleBindingLister) List() ([]ClusterRoleBinding, error) {
+	crbList, err := v1beta1.client.List(metav1.ListOptions{})
+	if err != nil {
+		return nil, errors.WithStack(err)
+	}
+	var crbs []ClusterRoleBinding
+	for _, crb := range crbList.Items {
+		crbs = append(crbs, v1beta1clusterrolebinding{crb: crb})
+	}
+
+	return crbs, nil
+}
+
+// NewClusterRoleBindingLister wraps the v1beta1 and v1 RBAC client's List with a common interface.
+func NewClusterRoleBindingLister(log logrus.FieldLogger, clientset kubernetes.Interface, discoveryHelper arkdiscovery.Helper) (ClusterRoleBindingLister, error) {
```

Comment:
We could change the signature here so instead of taking a `clientset`, it takes a `map[string]ListerConstructor`, where the keys are RBAC API versions, and the values are functions that return a `ClusterRoleBindingLister`. Then, instead of having a switch statement on concrete versions, we just do a lookup into the map for the server's preferred version.

This logic would be easily testable without actually requiring a `kubernetes.Interface`, and the correctness of the map (which would be defined in `plugin.go` could be determined by inspection.

Food for thought

Reply:
I like that idea - are you thinking of pairing that with a no-op lister then?

Reply:
Yeah, I think that'd be the way to go

Reply:
Sketching this out, it might look something like

```
// In rbac.go
type ListerConstructor func() ClusterRoleBindingLister
NewV1clusterrolebindinglister()
NewV1beta1clusterrolebindinglister()
NewRBACDisabledLister

// In plugins.go
listerFuncs := map[string]ListerConstructor {
rbac.SchemeGroupVersion.Version: NewV1clusterrolebindinglister,
rbacbeta.SchemeGroupVersion.Version: NewV1betaclusterrolebindinglister,
"": NewRBACDisabledLister // returns an empty list
}

action, err := NewServiceAccountAction(log, discoveryHelper, listerFuncs)

// In service_item_action.go/NewServiceAccountAction
// apigroup logic moved from NewClusterRoleBindingLister

crbLister := listerFuncs[apigroup.Version]()
crbs, err := crbLister.List()
if err != nil {
 return nil, err
}
return &serviceAccountAction{
log: log,
clusterRoleBindings: crbs,
```

I think it works; the listerFuncs map may also be specified in rbac.go and exported, looking at it here. WDYT?

Reply:
yep, looks good!